### PR TITLE
Mesh 518/Offchain worker

### PR DESCRIPTION
**Pull request by satyamakgec**
Offchain worker code is based on the latest substrate commit i.e - `3cc0973aff3f5b507c8ed1a13934bd35697b251f` I choose this because a lot of signing functionality was not present in the current version of the substrate so some of the code is commented to support compilation with the current version of the substrate.

### CODE REVIEW

**Comment by satyamakgec** [comment]
TODO:
- [ ] Filter the transaction created by the validators. eg. If the chain has 10 validators then 10 txn will be created after every block/cooling period so what will be the best to filter them and only execute one out of it.

File: pallets/cdd-offchain-worker/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+        // Firstly let's check that we call the right function.
+        if let Call::take_off_invalidate_nominators(block_number, target, signature) = call {
+            // Now let's check if the transaction has any chance to succeed.
+            let last_unsigned_at = <LastExtSubmittedAt<T>>::get();
+            if last_unsigned_at + T::CoolingInterval::get() > *block_number {
+                return InvalidTransaction::Stale.into();
+            }
+            // Let's make sure to reject transactions from the future.
+            let current_block = <system::Module<T>>::block_number();
+            if &current_block < block_number {
```
Comment (fmiguelgarcia) [inline pallets/cdd-offchain-worker/src/lib.rs:212]: `ensure!` pattern.

File: pallets/cdd-offchain-worker/src/lib.rs [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+                return InvalidTransaction::Future.into();
+            }
+
+            // Get the key from the store to verify whether the signed signature is
+            // generated from the validator or not.
+            let signer_id = match T::SignerId::all().into_iter().nth(0) {
```
Comment (fmiguelgarcia) [inline pallets/cdd-offchain-worker/src/lib.rs:218]: Replace `nth(0)` by `next()`, and use `ok_or`/`ok_or_else` pattern.

File: pallets/cdd-offchain-worker/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+                Some(id) => id,
+                None => return InvalidTransaction::BadProof.into(),
+            };
+
+            let signature_valid = target.using_encoded(|encoded_nominators| {
+                signer_id.verify(&encoded_nominators, &signature)
+            });
+
+            if !signature_valid {
```
Comment (fmiguelgarcia) [inline pallets/cdd-offchain-worker/src/lib.rs:227]: `ensure!` pattern.

File: pallets/cdd-offchain-worker/src/lib.rs [hunk 4]
```diff
@@ -0,0 +1,251 @@
...
+impl<T: Trait> Module<T> {
+    fn remove_invalidate_nominators(
+        block: T::BlockNumber,
+        invalid_nominators: Vec<T::AccountId>,
+    ) -> Result<(), &'static str> {
+        debug::native::info!("Enter into remove invalidate nominators functions");
+        use frame_system::offchain::SubmitUnsignedTransaction;
+        // First we validate whether the transaction proposer is validator or not.
+        // if yes then only the transaction get proposed otherwise not.
+        if !sp_io::offchain::is_validator() {
```
[the remark this answers is not in the document]
Comment (satyamakgec) [reply pallets/cdd-offchain-worker/src/lib.rs:155]: `ensure!` pattern can't be used in the `offchain_worker` call as its return type is `Result<(), ()>`

File: pallets/cdd-offchain-worker/src/lib.rs [hunk 5]
```diff
@@ -0,0 +1,246 @@
...
+
+            // Let's make sure to reject transactions from the future.
+            ensure!(
+                &<system::Module<T>>::block_number() >= block_number,
+                InvalidTransaction::Future
+            );
+
+            // Get the key from the store to verify whether the signed signature is
+            // generated from the validator or not.
+            let signer_id = T::SignerId::all()
```
Comment (maxsam4) [inline pallets/cdd-offchain-worker/src/lib.rs:215]: Where are these `SignerId` coming from? I don't see how the validator id is being set in this type/var.

File: pallets/staking/src/lib.rs [hunk 6]
```diff
@@ -1734,6 +1734,54 @@ decl_module! {
...
+                return None;
+            })
+            .collect::<Vec<T::AccountId>>();
+        return invalid_nominators;
+    }
+
+    /// POLYMESH-NOTE: This is Polymesh specific change
+    /// Here we are assuming that passed targets are always be a those nominators whose cdd
+    /// claim get expired or going to expire after the `buffer_time`.
+    pub fn unsafe_validate_cdd_expiry_nominators(targets: Vec<T::AccountId>) -> DispatchResult {
```
Comment (maxsam4) [inline pallets/staking/src/lib.rs:1766]: This gives validators the power to unilaterally remove all nominators (and validators) from the pool. This function should make sure that only nominators with expired cdd are being removed.

File: pallets/cdd-offchain-worker/src/lib.rs [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+            <LastExtSubmittedAt<T>>::put(current_block);
+            Self::deposit_event(RawEvent::InvalidateNominators(
+                <frame_system::Module<T>>::block_number(),
+                target,
+            ));
+            Ok(())
+        }
+
+
+        fn offchain_worker(block: T::BlockNumber) {
```
Comment (maxsam4) [inline pallets/cdd-offchain-worker/src/lib.rs:117]: Where is it configured how often this function should be called?  How is this function being automatically called?

File: pallets/cdd-offchain-worker/src/lib.rs [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+            // Print the debug statement to know that offchain worker initiated
+            debug::native::info!("Hello World from offchain workers!");
+            if Self::is_unsigned_transaction_allowed(block) {
+                debug::native::info!("Yeah transaction is allowed!");
+                // Fetch all the nominators whose cdd claim get expired or expiry remaining time is less
+                // than the `BufferInterval`. List of the Vec<Stash> get retrieved from the staking module
+                let invalid_nominators = <pallet_staking::Module<T>>::fetch_invalid_cdd_nominators((T::BufferInterval::get()).saturated_into::<u64>());
```
Comment (maxsam4) [inline pallets/cdd-offchain-worker/src/lib.rs:124]: I think this function should have a maximum bound (100?) or else this call can become so heavy that the chain gets stalled.

**Pull request merged**